stablehlo/dialect/StablehloOps.cpp

```diff
@@ -2294,148 +2294,112 @@ LogicalResult AllGatherOp::verify() {
+                             "same as the feature count,"
+                             " but the size of scale factor is "
+                          << scaleShape << " and the feature count is "
+                          << featureCount << ".";
+
+  return success();
+}
+
+// Refer ODS for properties that are already enforced including shapes and
+// element types. This verifier includes additional checks.
+LogicalResult BatchNormGradOp::verify() {
+  if (failed(verifyBatchNorm(getLoc(), operand(), feature_index(), scale())))
+    return failure();
+  return success();
+}
+
+LogicalResult BatchNormGradOp::inferReturnTypeComponents(
+    MLIRContext* context, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  BatchNormGradOp::Adaptor adaptor(operands, attributes, regions);
+
+  RankedTensorType operandType =
+      adaptor.operand().getType().cast<RankedTensorType>();
+  inferredReturnShapes.emplace_back(operandType.getShape(),
+                                    operandType.getElementType());
 
+  const int64_t featureCount = operandType.getShape()[adaptor.feature_index()];
+  SmallVector<int64_t> featureShape(1, featureCount);
+  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
+  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
   return success();
 }
 
 //===----------------------------------------------------------------------===//
 // BatchNormTrainingOp
 //===----------------------------------------------------------------------===//
 
+// Refer ODS for properties that are already enforced including shapes and
+// element types. This verifier includes additional checks.
 LogicalResult BatchNormTrainingOp::verify() {
-  // The following properties are already enforced by the ODS:
-  //  1. 'operand' and 'output' are ranked tensors.
-  //  2. 'scale', 'offset', 'batch_mean', 'batch_var' are 1D tensors.
-  //  3. Types of 'operand' and 'output' matches.
-  //  4. Same element-types for 'operand', 'batch_mean', & 'batch_var'.
-  //  5. Same shapes for 'scale', 'offset', 'batch_mean', & 'batch_var'.
-
-  auto operandType = operand().getType().cast<RankedTensorType>();
-  if (static_cast<int64_t>(feature_index()) >= operandType.getRank())
-    return emitOpError() << "expects feature_index to be smaller "
-                            "than the rank of operand type; got feature_index "
-                         << feature_index() << ", and rank "
-                         << operandType.getRank() << ".";
-
-  if (static_cast<int64_t>(feature_index()) < 0)
-    return emitOpError() << "expects feature_index to be a "
-                         << "non-negative number, got "
-                         << static_cast<int64_t>(feature_index()) << ".";
+  if (failed(verifyBatchNorm(getLoc(), operand(), feature_index(), scale())))
+    return failure();
```

Comment:
Are we checking that `operand`'s rank is at least 1? https://github.com/tensorflow/tensorflow/blob/d078f90b2833df402ff5c39e328f46e0bfd761b1/tensorflow/compiler/xla/service/shape_inference.cc#L1241-L1246

Reply:
There is an old comment I should add (bring back): `// Note:A valid value of feature-index implies 'operand_type.getRank() >=1'.` 
So, after we assert `0 <= feature_index < operandType.getRank()` in `verifyBatchNorm`, it implies that operand's rank is at least 1